hadoop-cloud/pom.xml

```diff
@@ -177,6 +214,188 @@
+          <artifactId>jackson-databind</artifactId>
+          <scope>${hadoop.deps.scope}</scope>
+        </dependency>
+        <dependency>
+          <groupId>com.fasterxml.jackson.core</groupId>
+          <artifactId>jackson-annotations</artifactId>
+          <scope>${hadoop.deps.scope}</scope>
+        </dependency>
+        <dependency>
+          <groupId>com.fasterxml.jackson.dataformat</groupId>
+          <artifactId>jackson-dataformat-cbor</artifactId>
+          <version>${fasterxml.jackson.version}</version>
+        </dependency>
+        <!--Explicit declaration to force in Spark version into transitive dependencies -->
+        <dependency>
+          <groupId>org.apache.httpcomponents</groupId>
+          <artifactId>httpclient</artifactId>
+          <scope>${hadoop.deps.scope}</scope>
+        </dependency>
+        <!--Explicit declaration to force in Spark version into transitive dependencies -->
+        <dependency>
+          <groupId>org.apache.httpcomponents</groupId>
+          <artifactId>httpcore</artifactId>
+          <scope>${hadoop.deps.scope}</scope>
+        </dependency>
+      </dependencies>
+    </profile>
+
+    <!--
+     Hadoop 3 simplifies the classpath, and adds a new committer base class which
+     enables store-specific committers.
+    -->
+    <profile>
+      <id>hadoop-3</id>
+      <properties>
+        <extra.source.dir>src/hadoop-3/main/scala</extra.source.dir>
+        <extra.testsource.dir>src/hadoop-3/test/scala</extra.testsource.dir>
+      </properties>
+
+      <build>
+        <plugins>
+          <!-- Include a source dir depending on the Scala version -->
+          <plugin>
+            <groupId>org.codehaus.mojo</groupId>
+            <artifactId>build-helper-maven-plugin</artifactId>
+            <executions>
+              <execution>
+                <id>add-scala-sources</id>
+                <phase>generate-sources</phase>
+                <goals>
+                  <goal>add-source</goal>
+                </goals>
+                <configuration>
+                  <sources>
+                    <source>${extra.source.dir}</source>
+                  </sources>
+                </configuration>
+              </execution>
+              <execution>
+                <id>add-scala-test-sources</id>
+                <phase>generate-test-sources</phase>
+                <goals>
+                  <goal>add-test-source</goal>
+                </goals>
+                <configuration>
+                  <sources>
+                    <source>${extra.testsource.dir}</source>
+                  </sources>
+                </configuration>
+              </execution>
+            </executions>
+          </plugin>
+        </plugins>
+
+      </build>
+      <dependencies>
+
+        <!--
+        There's now a hadoop-cloud-storage which transitively pulls in the store JARs,
+        but it still needs some selective exclusion across versions, especially 3.0.x.
```

Comment:
Can you expand a little on why the exclusions are needed? Some look a bit suspicious.

e.g. hadoop-common is already pulled transitively by other parts of Spark (including this very module, which does so via hadoop-client), so I'm not sure why the explicit exclusion is needed.

Reply:
Excluding hadoop-client means there's no need to worry about any of the stuff explicitly excluded from hadoop-client in the spark root pom (asm/asm, jackson, etc).

Hadoop 3.0.1 declares hadoop-client as a compile time dependency of [hadoop-cloud-storage](https://github.com/apache/hadoop/blob/branch-3.0.1/hadoop-cloud-storage-project/hadoop-cloud-storage/pom.xml)

From 3.0.2+ it's been cut down to provided, and added `azure-datalake` as a dependency [commit 3c03672e](https://github.com/apache/hadoop/commit/3c03672e876ddbd6a6425ea1a056ad13adc309ea), so it's complete w.r.t ASF connectors.
There's also a fix for the aws shaded SDK to exclude netty [HADOOP-15264](https://github.com/apache/hadoop/commit/e015e009897e481edc79f4ba72e2c53610b178a3), because of [aws-sdk-java/issues/1488](https://github.com/aws/aws-sdk-java/issues/1488). 

The individual hadoop cloud modules (hadoop-aws, hadoop-azure, ...) have also downgraded hadoop-client to being provided, so if you pull in any of those, you will only get the extra artifacts needed to connect to the relevant cloud endpoint, and are expected to pull in the same hadoop-client version elsewhere for things to work.

Here's the dependency list for spark-hadoop-cloud and 3.0.2-SNAPSHOT; 3.1 will be the same unless there's a last minute update to one of the external SDKs or jetty.

```
[INFO] +- org.apache.hadoop:hadoop-cloud-storage:jar:3.0.2-SNAPSHOT:compile
[INFO] |  +- org.apache.hadoop:hadoop-aliyun:jar:3.0.2-SNAPSHOT:compile
[INFO] |  |  \- com.aliyun.oss:aliyun-sdk-oss:jar:2.8.3:compile
[INFO] |  |     \- org.jdom:jdom:jar:1.1:compile
[INFO] |  +- org.apache.hadoop:hadoop-aws:jar:3.0.2-SNAPSHOT:compile
[INFO] |  |  \- com.amazonaws:aws-java-sdk-bundle:jar:1.11.271:compile
[INFO] |  +- org.apache.hadoop:hadoop-azure:jar:3.0.2-SNAPSHOT:compile
[INFO] |  |  +- com.microsoft.azure:azure-storage:jar:5.4.0:compile
[INFO] |  |  |  \- com.microsoft.azure:azure-keyvault-core:jar:0.8.0:compile
[INFO] |  |  \- org.eclipse.jetty:jetty-util-ajax:jar:9.3.19.v20170502:compile
[INFO] |  +- org.apache.hadoop:hadoop-azure-datalake:jar:3.0.2-SNAPSHOT:compile
[INFO] |  |  \- com.microsoft.azure:azure-data-lake-store-sdk:jar:2.2.5:compile
[INFO] |  \- org.apache.hadoop:hadoop-openstack:jar:3.0.2-SNAPSHOT:compile
```

Given that Hadoop 3.0.2+ is downgrading hadoop-client to provided, and that's the minimum version this patch will build against, then the exclusion is mostly superfluous: there to block regressions than actually keep it out.

Reply:
Update, tried taking it out. Turns out that the hadoop-allyun module still declares hadoop-common as provided, so it gets. in. The exclusion is needed, and I've filed [HADOOP-15354](https://issues.apache.org/jira/browse/HADOOP-15354). I don't know if that'll get into Hadoop 3.1.0 tho'